thirdparty/patchmatchnet/models/patchmatch.py

```diff
@@ -0,0 +1,965 @@
+                )
+                inverse_depth_interval = (inverse_min_depth - inverse_max_depth) * depth_interval_scale
+                inverse_depth_interval = inverse_depth_interval.view(batch_size, 1, 1, 1)
+
+                depth_sample = 1.0 / depth.detach() + inverse_depth_interval * depth_sample
+
+                depth_clamped = []
+                del depth
+                for k in range(batch_size):
+                    depth_clamped.append(
+                        torch.clamp(depth_sample[k], min=inverse_max_depth[k], max=inverse_min_depth[k]).unsqueeze(0)
+                    )
+
+                depth_sample = 1.0 / torch.cat(depth_clamped, dim=0)
+                del depth_clamped
+
+                return depth_sample
+
+
+class Propagation(nn.Module):
+    """ Propagation module implementation"""
+
+    def __init__(self, neighbors: int = 16) -> None:
+        """Initialize method
+
+        Args:
+            neighbors: number of neighbor views
+        """
+        super(Propagation, self).__init__()
+        self.neighbors = neighbors
+
+    def forward(
+        self,
+        batch: int,
+        height: int,
+        width: int,
+        depth_sample: torch.Tensor,
+        grid: torch.Tensor,
+        depth_min: int,
+        depth_max: int,
+        depth_interval_scale: float,
+    ) -> torch.Tensor:
+        # [B,D,H,W]
+        """Forward method of adaptive propagation
+
+        Args:
+            batch: batch size,
+            height: depth map height,
+            width: depth map width,
+            depth_sample: sample depth map, in shape of [batch, num_depth, height, width],
+            grid: 2D grid for bilinear gridding, in shape of [batch, len(original_offset)*H, W, 2],
+                    len(original_offset) generally equals to 9, suggesting the 3x3 neighbor grids.
+            depth_min: minimum virtual depth,
+            depth_max: maximum virtual depth,
+            depth_interval_scale: depth virtual interval scale,
+
+        Returns:
+            propagate depth: sorted propagate depth map [batch, num_depth+num_neighbors, height, width]
+        """
+        num_depth = depth_sample.size()[1]
+        propagate_depth = depth_sample.new_empty(batch, num_depth + self.neighbors, height, width)
+        propagate_depth[:, 0:num_depth, :, :] = depth_sample
+
+        propagate_depth_sample = F.grid_sample(
+            depth_sample[:, num_depth // 2, :, :].unsqueeze(1), grid, mode="bilinear", padding_mode="border"
+        )
+        del grid
+        propagate_depth_sample = propagate_depth_sample.view(batch, self.neighbors, height, width)
+
+        propagate_depth[:, num_depth:, :, :] = propagate_depth_sample
+        del propagate_depth_sample
+
+        # sort
+        propagate_depth, _ = torch.sort(propagate_depth, dim=1)
+
+        return propagate_depth
+
+
+class Evaluation(nn.Module):
+    """Evaluation Class for adaptive evaluation step in Learning-based Patchmatch

```

Comment:
I might call it "Evaluation module" instead of class. This conveys the semantics, instead of just implementation